types/yup/index.d.ts

```diff
@@ -58,12 +59,12 @@ export interface Schema<T> {
     meta(): any;
     describe(): SchemaDescription;
     concat(schema: this): this;
-    validate(value: T, options?: ValidateOptions): Promise<T>;
-    validateSync(value: T, options?: ValidateOptions): T;
-    validateAt(path: string, value: T, options?: ValidateOptions): Promise<T>;
-    validateSyncAt(path: string, value: T, options?: ValidateOptions): T;
-    isValid(value: T, options?: any): Promise<boolean>;
-    isValidSync(value: T, options?: any): boolean;
+    validate(value: any, options?: ValidateOptions): Promise<T>;
+    validateSync(value: any, options?: ValidateOptions): T;
+    validateAt(path: string, value: any, options?: ValidateOptions): Promise<T>;
+    validateSyncAt(path: string, value: any, options?: ValidateOptions): T;
+    isValid(value: any, options?: any): Promise<boolean>;
+    isValidSync(value: any, options?: any): boolean;
```

Comment:
I'm a bit uncomfortable with the amount of `any`s here, doesn't this lose type inference on the returned value?

Reply:
Good point. I think this can be tweaked. For context, here is my note in the PR description:

> All flavors of Schema<T>.validate*() were changed to test arguments of type `any` rather than `T`, because their purpose is to validate an object of potentially unknown type and origin (e.g. parsed from JSON) against the schema.

To specifically address your question, I believe that `validate()`, `validateSync()`, `isValid()`, and `isValidSync()` should be changed. Here's the reasoning: The returned value is unaffected by this change and overall type safety is preserved. This is because `T` is still returned in the first two methods **if and only if the `value` argument passes the schema rules, otherwise an exception is thrown**. This simply allows for validation of `value`s whose types are not known a priori (like in the JSON case mentioned above). I believe this is good thing because it will prevent developers from having to cast such `value` arguments to type `T` before passing them to these methods.

But perhaps the `isValid()` and `isValidSync()` methods' return types could be enhanced from `boolean` to `value is T`.

But you're right to question the changes to `validateAt()` and `validateSyncAt()`. I'm on the fence about it. Do you need a well-typed object graph in order to validate a particular subbranch of it?

Reply:
To answer the question right above this post about having a well typed object in order to validate a particular sub branch, I think we need to make a judgement call here. If we accept `any` as input, then we're saying, you can use this to validate anything that takes the form of `Partial<T>` which might be useful in the case where I have something else that's independent from the schema but still follows the same validation rules. For example, a validation schema for firstname, email and address being able to validate an email field coming from somewhere else in the form. 

On the other hand, if we don't accept `any` and we force `T`, it could contribute to early detection that this object does not look like something that would pass the validation schema which would tell the developer that they either need to transform the object to something that the validation schema can accept or they have passed in the wrong object to be validated.

In my mind, the first scenario in this post is more contrived than the second, so I am leaning towards keeping `T` as the type for `validateAt()` and `validateSyncAt()` but I also recognize that `any` is worthwhile. 

On the `validate()` and `validateAt()` side, I definitely agree that `any` should be the type since you want to know if input that you don't trust validates against what you expect it to be and secondly (as I learned a few weeks ago), it will also parse JSON which is pretty :100:

Reply:
Good points, I think it would be benefitial to accept `any` as an argument as the return value is typed as expected. 
Whichever way the hammer falls you can usually just use another variable to get the type-check or type cast.
eg.

Early warning:
```
const value: MyType = jsonBlob;
```

Type cast to accept blob as-is:
```
await schema.validate(jsonBlob as MyType)
```

I'm hesitant to change it as I don't know how other people are using this and if they rely on this kind of early warning. Secondly, at what point does this classify as a breaking change?